seagoat/sources/chroma.py

```diff
@@ -87,28 +142,138 @@ def fetch(query_text: str, limit: int):
-                documents=[chunk.chunk],
-                metadatas=[
-                    {
+                ids=ids,
+                documents=documents,
+                metadatas=metadatas,
+            )
+        except IDAlreadyExistsError as e:
+            # If batch fails due to duplicate IDs, try individually
+            logger.debug(f"Batch add failed with duplicate IDs, retrying individually: {e}")
+            for i, (chunk_id, doc, meta) in enumerate(zip(ids, documents, metadatas)):
+                try:
+                    chroma_collection.add(
+                        ids=[chunk_id],
+                        documents=[doc],
+                        metadatas=[meta],
+                    )
+                except IDAlreadyExistsError:
+                    # This specific chunk already exists, skip it
+                    logger.debug(f"Chunk {chunk_id} already exists, skipping")
+                except Exception as e:
+                    # Other error for this chunk
+                    logger.error(f"Failed to add chunk {chunk_id}: {e}")
+                    failed_indices.append(i)
+        except Exception as e:
+            # Other batch error - try to identify which chunks are problematic
+            logger.error(f"Batch add failed: {e}")
+            # Try each chunk individually to identify failures
+            for i, (chunk_id, doc, meta) in enumerate(zip(ids, documents, metadatas)):
+                try:
+                    chroma_collection.add(
+                        ids=[chunk_id],
+                        documents=[doc],
+                        metadatas=[meta],
+                    )
+                except Exception as e:
+                    logger.debug(f"Failed to add chunk {chunk_id}: {e}")
+                    failed_indices.append(i)
+        
+        return failed_indices
+
+    def flush_batch():
+        """Flush the current batch to ChromaDB with improved error handling"""
+        if not batch_buffer or batch_buffer.is_empty():
+            return
+        
+        # Get current batch data
+        ids = batch_buffer.ids.copy()
+        documents = batch_buffer.documents.copy()
+        metadatas = batch_buffer.metadatas.copy()
+        
+        # Clear buffer immediately to avoid re-processing
+        batch_buffer.clear()
+        
+        # Try to add chunks
+        failed_indices = _add_chunks_to_collection(ids, documents, metadatas)
+        
+        # Store failed chunks for potential retry
+        for idx in failed_indices:
+            failed_chunks.append({
+                'id': ids[idx],
+                'document': documents[idx],
+                'metadata': metadatas[idx]
+            })
+        
+        if failed_indices:
+            logger.warning(f"Failed to add {len(failed_indices)} chunks in batch")
+
+    def cache_chunk(chunk):
+        if not batch_buffer:
+            # No batching - use original behavior with error handling
```

Comment:
could we simply reuse the batching behavior here, but with a batch size of 1?

would that have any downsides?

if we do that, we could remove the old logic completely